test/server.go

```diff
@@ -0,0 +1,252 @@
+
+	// CheckAttributes is the channel on which the attribute bag generated by
+	// the server for Check() requests is sent.
+	CheckAttributes chan attribute.Bag
+
+	// QuotaDispatches is the channel on which the quota information generated by
+	// the server for Check() requests is sent.
+	QuotaDispatches chan QuotaDispatchInfo
+
+	// ReportAttributes is the channel on which the attribute bag generated by
+	// the server for Report() requests is sent.
+	ReportAttributes chan attribute.Bag
+}
+
+// NewAttributesServer creates an AttributesServer. All channels are set to
+// default length.
+func NewAttributesServer() *AttributesServer {
+	list := attribute.GlobalList()
+	globalDict := make(map[string]int32, len(list))
+	for i := 0; i < len(list); i++ {
+		globalDict[list[i]] = int32(i)
+	}
+
+	return &AttributesServer{
+		globalDict,
+		false,
+		nil,
+		true,
+		make(chan attribute.Bag),
+		make(chan QuotaDispatchInfo),
+		make(chan attribute.Bag),
+	}
+
+}
+
+// Check sends a copy of the protocol buffers attributes wrapper for the preconditions
+// check as well as for each quotas check to the CheckAttributes channel. It also
+// builds a CheckResponse based on server fields. All channel sends timeout to
+// prevent problematic tests from blocking indefinitely.
+func (a *AttributesServer) Check(ctx context.Context, req *mixerpb.CheckRequest) (*mixerpb.CheckResponse, error) {
+
+	if a.GenerateGRPCError {
+		return nil, errors.New("error handling check call")
+	}
+
+	requestBag := attribute.NewProtoBag(&req.Attributes, a.GlobalDict, attribute.GlobalList())
+	defer requestBag.Done()
+
+	if a.PublishToChannels {
+		// avoid blocked go-routines in testing
+		select {
+		case a.CheckAttributes <- attribute.CopyBag(requestBag): // no-op
+		case <-time.After(3 * time.Second):
+			return nil, errors.New("timeout trying to send attribute bag")
+		}
+	}
+
+	resp := &mixerpb.CheckResponse{
+		Precondition: mixerpb.CheckResponse_PreconditionResult{
+			Status:               rpc.Status{Code: int32(rpc.OK)},
+			ValidUseCount:        DefaultValidUseCount,
+			Attributes:           mixerpb.Attributes{},
+			ReferencedAttributes: mixerpb.ReferencedAttributes{},
+		},
+	}
+
+	if len(req.Quotas) > 0 {
+		resp.Quotas = make(map[string]mixerpb.CheckResponse_QuotaResult, len(req.Quotas))
+		for name, param := range req.Quotas {
+			qma := QuotaMethodArgs{
+				Quota:           name,
+				Amount:          param.Amount,
+				DeduplicationID: req.DeduplicationId + name,
+				BestEffort:      param.BestEffort,
+			}
+
+			if a.PublishToChannels {
+				// avoid blocked go-routines in testing
+				select {
+				case a.QuotaDispatches <- QuotaDispatchInfo{attribute.CopyBag(requestBag), qma}: //no-op
+				case <-time.After(3 * time.Second):
+					return nil, errors.New("timeout trying to send quota dispatch")
+				}
+			}
+
+			qr := mixerpb.CheckResponse_QuotaResult{
+				GrantedAmount:        param.Amount,
+				ValidDuration:        DefaultValidDuration,
```

Comment:
good news: you can. If you set the `CheckResponse` field in the `AttributesServer`, then all of this will be ignored and overridden.

```
attrSrv.CheckResponse = &mixerpb.CheckResponse{ ... Quotas: ... "quota1": mixerpb.CheckResponse_QuotaResult{ GrantedAmount: myCustomAmt, ... } }
...
client.Check(...)
```